consensus/parlia/snapshot.go

```diff
@@ -289,18 +342,22 @@ func (s *Snapshot) apply(headers []*types.Header, chain consensus.ChainHeaderRea
 					}
 				}
 			}
-			oldLimit := len(snap.Validators)/2 + 1
-			newLimit := len(newVals)/2 + 1
-			if newLimit < oldLimit {
-				for i := 0; i < oldLimit-newLimit; i++ {
-					delete(snap.Recents, number-uint64(newLimit)-uint64(i))
+			if chainConfig.IsBohr(header.Number, header.Time) {
+				epochClearKey := math.MaxUint64 - header.Number.Uint64()/s.config.Epoch // impossible used as a block number
+				// in a epoch, after the first validator set switch, minerHistoryCheckLen() may become larger,
+				// so the unexpected second switch will happen, don't clear up the `Recents` in this kind of scene.
+				if _, ok := snap.Recents[epochClearKey]; !ok {
+					snap.Recents = make(map[uint64]common.Address) // without this logic, there will be several off-turn blocks when do validator set switch
+					snap.Recents[epochClearKey] = common.Address{}
+					log.Debug("Recents are cleared up", "blockNumber", number)
 				}
-			}
-			oldLimit = len(snap.Validators)
-			newLimit = len(newVals)
-			if newLimit < oldLimit {
-				for i := 0; i < oldLimit-newLimit; i++ {
-					delete(snap.RecentForkHashes, number-uint64(newLimit)-uint64(i))
+			} else {
+				oldLimit := len(snap.Validators)/2 + 1
```

Comment:
seems not exact same with previous logic?

Reply:
> seems not exact same with previous logic?

same, the previous line is 292, not 299